cupy/cuda/cub.pyx

```diff
@@ -55,29 +57,32 @@ cdef extern from 'cupy_cub.h':
 # Python interface
 ###############################################################################
 
-cpdef _preprocess_array(ndarray arr, axis, bint keepdims):
+cpdef _preprocess_array(ndarray arr, axis, bint keepdims, str order):
     '''
     This function more or less follows the logic of _get_permuted_args() in
-    reduction.pxi. The input array arr is C-contiguous along axis.
+    reduction.pxi. The input array arr is C- or F- contiguous along axis.
     '''
     # if import at the top level, a segfault would happen when import cupy!
     from cupy.core._kernel import _get_axis
 
     cdef tuple reduce_axis, out_axis, axis_permutes, out_shape
     cdef Py_ssize_t contiguous_size = 1
+    cdef list temp = []
 
     reduce_axis, out_axis = _get_axis(axis, arr._shape.size())
-    axis_permutes = out_axis + reduce_axis
     # one more sanity check?
+    if order == 'C':
+        axis_permutes = out_axis + reduce_axis
+    elif order == 'F':
+        axis_permutes = reduce_axis + out_axis
     if axis_permutes != tuple(range(len(arr.shape))):
-        raise ValueError("should not happen")
+        raise ValueError('should not happen')
 
     for axis in reduce_axis:
         contiguous_size *= arr.shape[axis]
     if not keepdims:
         out_shape = tuple([arr.shape[axis] for axis in out_axis])
     else:
-        temp = []
         for axis in range(arr.ndim):
             if axis in out_axis:
                 temp.append(arr.shape[axis])
```

Comment:
Instead of using a temp variable, how about 
```
out_shape = tuple((arr.shape[axis] if axis in out_axis else 1 for axis in range(arr.ndim)))

Reply:
Sure, let me try to break the long line in a nice way...

Reply:
Fixed in 9cc059f.